### ISSUE TITLE
Strings: Fallback to development language when key is missing

### ISSUE THREAD

**Issue by chrisjrex** [SwiftGen/SwiftGen#937]
Creating a new issue since #381 was closed. - I believe I have addressed the concerns of previous reviewers and provided more clarity on the source of the issue.

swiftgen --version = **6.5.1**

Happy to create the PR, once there is some consensus. 

## The Problem

Xcode does not handle fallback for partially translated .strings files, only when the file is missing. To state the issue more clearly, ill take a paragraph from this article that explains it well.

[Article link](https://samwize.com/2018/01/23/localization-fall-back-to-base-language/)

<img width="700" alt="CleanShot 2022-05-09 at 14 21 57@2x" src="https://user-images.githubusercontent.com/4457170/167419193-dca151a0-fdea-4aa6-b025-61749645c837.png">

> **Fallback 3: Unsupported Phrase**
Let’s take the example where a user prefers zh-Hans.
>
>What happens when iOS try to get the NSLocalizedString of `“another-title”`?
Remember, zh-Hans is not fully translated. It does not have `“another-title”`. What do you think will happen?
>
>1) Fallback to development language en, or
>2) return `“another-title”`
>
>Many developers think it is (1). Unfortunately, it is (2), with that ugly key name!

## Solution

Here is an updated version of the code by @jonandersen in 381, that handles any development language as the base/fallback.

```
func localizeStringOrFallback(_ table: String, _ key: String, _ args: CVarArg...) -> String {
    let bundle = BundleToken.bundle
    let format = NSLocalizedString(key, tableName: table, bundle: bundle, comment: "")
    let string = String(format: format, locale: Locale.current, arguments: args)

    // If string key does not find a match within the Locale.current file. Look up the fallback string.
    // The edge case of this is that if a key and value are expected to match: "Foo" = "Foo", it will trigger the fallback logic.
    if string == key,
       let developmentRegion = bundle.infoDictionary?["CFBundleDevelopmentRegion"] as? String,
       let path = bundle.path(forResource: developmentRegion, ofType: "lproj"),
       let bundle = Bundle(path: path)
    {
        let format = NSLocalizedString(key, bundle: bundle, comment: "")
        return String(format: format, locale: Locale(identifier: developmentRegion), arguments: args)
    } else {
        return string
    }
}
```

I have tested this solution, and created a custom stencil implementing this logic. However, this complicates maintenance of the repo when SwiftGen releases updated stencils for new Swift versions.

I also think it would be a useful solution to others. Where showing fallback text is likely a better "bug" than displaying a key.

## Known issue with this approach
One known issue with this is that if a key and value are expected to match: "Foo" = "Foo", it will trigger the fallback logic.

Therefore if you had:
```
// en.proj
"Foo" = "A string that does not match its key";

// fr.proj
"Foo" = "Foo";      <------ if there was a use-case to repeat the key.
```
In this case, the displayed string would be "A string that does not match its key".

## Stencil Code

```
extension {{enumName}} {
  private static func tr(_ table: String, _ key: String, _ args: CVarArg...) -> String {
    return localizeStringOrFallback(table, key, args)
  }
}

extension {{enumName}} {
  func localizeStringOrFallback(_ table: String, _ key: String, _ args: CVarArg...) -> String {
    let bundle = BundleToken.bundle
    let format = NSLocalizedString(key, tableName: table, bundle: bundle, comment: "")
    let string = String(format: format, locale: Locale.current, arguments: args)

    // If string key does not find a match within the Locale.current file. Look up the fallback string.
    // The edge case of this is that if a key and value are expected to match: "Foo" = "Foo", it will trigger the fallback logic.
    if string == key,
       let developmentRegion = bundle.infoDictionary?["CFBundleDevelopmentRegion"] as? String,
       let path = bundle.path(forResource: developmentRegion, ofType: "lproj"),
       let bundle = Bundle(path: path)
    {
        let format = NSLocalizedString(key, bundle: bundle, comment: "")
        return String(format: format, locale: Locale(identifier: developmentRegion), arguments: args)
    } else {
        return string
    }
  }
}
```

## Alternative solution

Don't change anything, and expect dev's to insert an entry for all locales, even if not translated. i.e. Insert the fallback string manually.

This is error prone, ive seen it in apps ive worked on. where keys might be added to `en` but not `en-US`. and then `en-US` users are seeing keys displayed in the UI as text.

**Comment by djbe** [comment]
The article weirdly enough ignores the built-in functionality of the function it's using:
`localizedString(forKey:value:table:)`

Quote from [docs](https://developer.apple.com/documentation/foundation/bundle/1417694-localizedstring):
> value
> The value to return if key is nil or if a localized string for key can’t be found in the table.

I've created a PR to implement the needed changes in our built-in templates: #964.

Note: use this instead of `NSLocalized`, the latter is meant for string file generation, and Xcode will complain about using non-constant strings as parameters for it.

**Comment by djbe** [comment]
Implemented by SwiftGen 6.6.0, which has been released.
